### PR TITLE
Remove extra forward pass from dynamic MSE loss

### PR DESCRIPTION
**Pull request by amogh-gulati**
Resolved this issue #443

Eliminates the redundant second model forward pass that was used solely to update DynamicLoss channel scales. Instead, we reuse the per-channel loss already computed during the training rollout.

This should introduce a delayed estimate (if i interpret it correctly)

We are also using the error from the end of the 4-step rollout rather than after a single step.

### CODE REVIEW

File: src/ocean_emulators/train.py [hunk 1]
```diff
@@ -595,16 +595,24 @@ def train_one_epoch(self, epoch):
         logger.info(f"Aggregating train logs")
         return train_aggregator.get_logs()
 
-    def _call_loss_update(self, data: TrainData):
-        # This is a separate function to ensure locals are dropped immediately after use
+    def _call_loss_update(self, loss_per_channel: torch.Tensor):
```
Comment (alxmrs) [inline src/ocean_emulators/train.py]: :sheep: consider making this accept the `TO` object instead of a raw tensor in case we need other data later / to add more context to the shape of the tensor.

File: src/ocean_emulators/train.py [hunk 2]
```diff
@@ -595,16 +595,24 @@ def train_one_epoch(self, epoch):
...
+        # Use the already-computed per-channel loss from the training rollout
+        # to update DynamicLoss scales, avoiding an second forward pass.
+        # This introduces delayed estimate but should be more efficient
```
Comment (alxmrs) [inline src/ocean_emulators/train.py]: I appreciate you noting the tradeoffs involved here. By delayed estimate, do we mean a delay in wandb logs?

File: src/ocean_emulators/train.py (hunk 1 above)
Comment (alxmrs) [reply src/ocean_emulators/train.py]: Can we change this name to `_maybe_call_loss_update` -- we don't take any action if there is no dynamic loss.

File: src/ocean_emulators/train.py [hunk 3]
```diff
@@ -595,16 +595,20 @@ def train_one_epoch(self, epoch):
...
-                input, label = data[0]
-                single_step_data.append(input, label)
-                pred = self.model(single_step_data)
-                update(pred[0], label, ctx=data.ctx)
+            # Undo the dynamic scaling to recover the raw per-channel loss.
+            if scale_fn := getattr(self.loss_fn, "loss_scale_per_channel", None):
+                scale = scale_fn()
+                raw_loss = (
+                    loss_per_channel.detach().reshape(-1, scale.shape[0]) / scale
+                ).reshape(-1)
```
Comment (alxmrs) [inline src/ocean_emulators/train.py]: 🐑 A few naming nits, to be more consistent with the dynamic loss class: 

Totally optional.
```suggestion
            if get_scales := getattr(self.loss_fn, "loss_scale_per_channel", None):
                per_channel_scale = get_scales()
                raw_loss = (
                    loss_per_channel.detach().reshape(-1, scale.shape[0]) / per_channel_scale
                ).reshape(-1)
```

File: src/ocean_emulators/train.py [hunk 4]
```diff
@@ -595,16 +595,20 @@ def train_one_epoch(self, epoch):
...
+            else:
+                raw_loss = loss_per_channel.detach()
```
Comment (alxmrs) [inline src/ocean_emulators/train.py]: I don't think we ever hit this case: the only function that has an `update` is the dynamic loss right now. Maybe, instead of a default behavior, we can throw an error -- this would alert us early that something is out of place (rather than silently getting strange results).

File: src/ocean_emulators/train.py (hunk 1 above)
Comment (alxmrs) [reply src/ocean_emulators/train.py]: Thinking about it more, I like the name `_maybe_update_loss` better.

**Review by alxmrs — COMMENTED** [review]
Thanks for doing this! Generally, this is looking good. A few notes.

File: src/ocean_emulators/train.py (hunk 1 above)
Comment (amogh-gulati) [reply src/ocean_emulators/train.py]: This makes sense

File: src/ocean_emulators/train.py (hunk 2 above)
Comment (amogh-gulati) [reply src/ocean_emulators/train.py]: by delayed estimate here we mean that the loss scales are one step older, we have already updated the model weights before we call this function. Earlier we were using the new weights to get the raw loss again which was more upto date with the weights ( but even that was only doing 1 step pass instead of 4 so that was not exactly precise either )

File: src/ocean_emulators/train.py (hunk 4 above)
Comment (amogh-gulati) [reply src/ocean_emulators/train.py]: yup error makes sense sense

**Comment by jder** [comment]
@amogh-gulati is this ready to merge? One thing that we might want to consider: this changes the semantics considerably (using the 4-step rollout which seems better, and delaying the update by 1 step which seems maybe worse but not enough to matter) but we might want to use the existing behavior for any other runs being done specifically for the paper. Could we add this as a new loss (or new configuration for the existing loss) until we ship the paper?

**Comment by amogh-gulati** [comment]
> @amogh-gulati is this ready to merge? One thing that we might want to consider: this changes the semantics considerably (using the 4-step rollout which seems better, and delaying the update by 1 step which seems maybe worse but not enough to matter) but we might want to use the existing behavior for any other runs being done specifically for the paper. Could we add this as a new loss (or new configuration for the existing loss) until we ship the paper?

This should be ready to merge but yes I agree, lets keep it as a configuration (maybe a bool for delayed estimate )

**Comment by alxmrs** [comment]
@amogh-gulati I see you marked this as a draft, is it ready for my review?

**Review by alxmrs — APPROVED** [review]
LGTM, thanks for the changes.

**Comment by amogh-gulati** [comment]
> LGTM, thanks for the changes.

wanted to support delayed estimate as a bool first 😅

**Comment by amogh-gulati** [comment]
> @amogh-gulati is this ready to merge? One thing that we might want to consider: this changes the semantics considerably (using the 4-step rollout which seems better, and delaying the update by 1 step which seems maybe worse but not enough to matter) but we might want to use the existing behavior for any other runs being done specifically for the paper. Could we add this as a new loss (or new configuration for the existing loss) until we ship the paper?

does the current change look good?

**Comment by jder** [comment]
@codex review

**Comment by jder** [comment]
@amogh-gulati thanks, this looks good to me now.

🐑 FWIW I think it would be fine to keep the old signature of `update` which takes pred/target and recomputes the (unscaled) loss rather than having to go through the unscaling and/or grabbing the underlying loss_fn from the dynamic loss to get unscaled loss externally. Would be slightly slower but I doubt meaningfully slower.

**Review by jder — APPROVED** [review]

**Comment by amogh-gulati** [comment]
> @amogh-gulati thanks, this looks good to me now.
> 
> 🐑 FWIW I think it would be fine to keep the old signature of `update` which takes pred/target and recomputes the (unscaled) loss rather than having to go through the unscaling and/or grabbing the underlying loss_fn from the dynamic loss to get unscaled loss externally. Would be slightly slower but I doubt meaningfully slower.

that does sound cleaner but i dont have access to preds in the delayed estimate path

**Comment by jder** [comment]
> > @amogh-gulati thanks, this looks good to me now.
> > 🐑 FWIW I think it would be fine to keep the old signature of `update` which takes pred/target and recomputes the (unscaled) loss rather than having to go through the unscaling and/or grabbing the underlying loss_fn from the dynamic loss to get unscaled loss externally. Would be slightly slower but I doubt meaningfully slower.
> 
> that does sound cleaner but i dont have access to preds in the delayed estimate path

Ah, right, because we still do loss calculation in the model. We really should fix that one of these days. Thanks for taking a look at it!
